### PR TITLE
Pull request for issue #521 https://github.com/inversify/InversifyJS/…

### PR DESCRIPTION
**Pull request by AltekkeE**
Add Decorators for binding method parameters to request parameters like in Spring or node-decorators

## Description
@Request - binds a method parameter to the request object - still provided at the end of the arguments list for backwards compatibility
@Response - binds a method parameter to the response object - still provided at the end of the arguments list for backwards compatibility
@RequestParam - binds a method parameter to request.params or to a specific parameter if a name is passed
@QueryParam - binds a method parameter to request.query or to a specific query parameter if a name is passed
@RequestBody - binds a method parameter to request.body or to a specific body property if a name is passed
@RequestHeaders - binds a method parameter to the request headers
@Cookies - binds a method parameter to the request cookies
@Next - binds a method parameter to the next() function - still provided at the end of the arguments list for backwards compatibility

## Related Issue
issue 521
https://github.com/inversify/InversifyJS/issues/521

## Motivation and Context
Better developer experience than using request and response
modeled after spring and node-decorators
some code reused from node-decorators

## How Has This Been Tested?
unit tests for new code
original unit tests still pass
backwards compatible
node v7.4.0 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -2,16 +2,16 @@
...
-		"src/**/*.js": true,
-		"test/**/*.js": true,
-        "**/es": true,
-        "**/lib": true,
-        "**/dts": true,
-        "**/coverage": true,
-        "**/dist": true,
-        "**/docs": true,
-		"type_definitions/**/*.js": true
+ 		"**/.DS_Store": true,
```
Comment (Dirrk) [inline .vscode/settings.json]: This looks like auto formatting was it intended to be committed?

File: src/decorators.ts [hunk 2]
```diff
@@ -1,10 +1,10 @@
 import * as express from "express";
 import { interfaces } from "./interfaces";
-import { METADATA_KEY } from "./constants";
+import { METADATA_KEY, ParameterType } from "./constants";
 
 export function Controller(path: string, ...middleware: interfaces.Middleware[]) {
     return function (target: any) {
-        let metadata: interfaces.ControllerMetadata = {path, middleware, target};
+        let metadata: interfaces.ControllerMetadata = { path, middleware, target };
```
Comment (Dirrk) [inline src/decorators.ts]: Again I think this is your ide making an unexpected change?

File: src/decorators.ts [hunk 3]
```diff
@@ -39,7 +39,7 @@ export function Delete(path: string, ...middleware: interfaces.Middleware[]): in
 
 export function Method(method: string, path: string, ...middleware: interfaces.Middleware[]): interfaces.HandlerDecorator {
     return function (target: any, key: string, value: any) {
-        let metadata: interfaces.ControllerMethodMetadata = {path, middleware, method, target, key};
+        let metadata: interfaces.ControllerMethodMetadata = { path, middleware, method, target, key };
```
Comment (Dirrk) [inline src/decorators.ts]: Formatting

File: src/constants.ts [hunk 4]
```diff
@@ -4,9 +4,21 @@ const TYPE = {
 
 const METADATA_KEY = {
     controller: "_controller",
-    controllerMethod: "_controller-method"
+    controllerMethod: "_controller-method",
+    controllerParameter: "_controller-parameter"
 };
 
+export enum ParameterType {
```
Comment (Dirrk) [inline src/constants.ts]: Should this be capitalized like the other constants?

**Comment by lholznagel** [comment]
Thanks!
Will also take a look later

**Comment by AltekkeE** [comment]
Thanks for pointing those things out. I noticed some other things that were
auto formatted as well. I was rushing because I had everything set up to
submit it and my kids got home and interrupted me. Should I resubmit the
pull request?

**Comment by lholznagel** [comment]
@AltekkeE just push more commits to the current branch. It will auto update this pr :)

**Comment by AltekkeE** [comment]
@lholznagel  - Done.

I also notice that the code coverage report is not working correctly. It says 100% but no files are actually checked. This may be happening in the main inversify repo too.

File: README.md [hunk 5]
```diff
@@ -42,12 +42,17 @@ export class FooController implements interfaces.Controller {
     constructor( @inject('FooService') private fooService: FooService ) {}
 
     @Get('/')
-    private index(req: express.Request): string {
+    private index(req: express.Request, res: express.Response, next: express.NextFunction): string {
```
Comment (lholznagel) [inline README.md]: Couldn´t we just use `@QueryParam('id') id: string`?

Comment (AltekkeE) [inline README.md]: I wanted to show that you can still do it the original way

Comment (AltekkeE) [inline README.md]: I wanted to show that you can still do it the original way

Comment (lholznagel) [inline README.md]: Oh okay.
Makes sense

**Comment by lholznagel** [comment]
LGTM :+1:

**Comment by remojansen** [comment]
:+1: Amazing PR 💯 thanks a lot for this contribution. One question: did you find a way to fix the test coverage reports?

**Comment by lholznagel** [comment]
@remojansen yesterday I checked travis and the problem is also in the main repo since some time.
I did a quick look to find a solution, but until now I didn´t find anything.

**Comment by remojansen** [comment]
I think is probably some breaking change in one of the tools but is hard to find out which one :(

**Comment by AltekkeE** [comment]
I looked at it briefly but I didn't know if you were doing it for
transpired code or for typescript. It could have to do with not having
source maps enabled. I had found some other people that ran into a similar
problem but forget if it was recent. They solved it by making a new gulp
plugin for typescript to Istanbul

**Comment by remojansen** [comment]
Thanks a lot I will create an issue to track this down and see if we can fix it :+1:

**Comment by remojansen** [comment]
Just release `inversify-express-utils@3.4.0` which includes this feature :tada:

**Comment by remojansen** [comment]
Hi @AltekkeE I have noticed that in the docs you mention `@RequestParams` but it seems to be not available in the source code. Did you forget to commit it? Thanks!

**Comment by AltekkeE** [comment]
@remojansen I will take a look right now

**Comment by AltekkeE** [comment]
:grimacing: Spelling mistake, Should be @RequestParam in the docs. It is correct down in the description part just not in that delete statement

**Comment by remojansen** [comment]
No worries.Would u like to send a PR?

**Comment by AltekkeE** [comment]
@remojansen [Gulp Problem] in a project that I am working on we are not using gulp. We just run the nyc using npm 
``` "scripts": {
    "start": "node ./dist/bootstrap.js",
    "postinstall": "tsc",
    "test": "mocha --opts mocha.opts && rm -rf test",
    "test:dev": "mocha --opts mocha.opts --watch",
    "cover": "nyc npm run test"
  }```
the issue below points to this as well.  

https://github.com/SBoudrias/gulp-istanbul/issues/115

**Pull request merged**